File: src/main/java/io/github/saneea/feature/text/Trim.java
package io.github.saneea.feature.text;

import java.io.IOException;
import java.io.Reader;
import java.io.Writer;
import java.util.LinkedList;
import java.util.Queue;
import java.util.function.Supplier;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.Option;

import io.github.saneea.Feature;
import io.github.saneea.FeatureContext;

public class Trim implements//
		Feature, //
		Feature.CLI, //
		Feature.In.Text.Reader, //
		Feature.Out.Text.Writer {

	private static final String LEFT_TRIM = "leftTrim";
	private static final String RIGHT_TRIM = "rightTrim";

	private Reader in;
	private Writer out;
	private CommandLine commandLine;

	@Override
	public String getShortDescription() {
		return "trim leading and/or trailing whitespaces";
	}

	@Override
	public void run(FeatureContext context) throws IOException {

		TextConverter converter = TextConverter.orig();

		converter = addTrimmer(LEFT_TRIM, converter, LeftTrimmer::new);
		converter = addTrimmer(RIGHT_TRIM, converter, RightTrimmer::new);

		int charCode;
		while ((charCode = in.read()) != -1) {
			converter.convertChar(charCode, out::write);
		}
	}

	private TextConverter addTrimmer(//
			String optName, //
			TextConverter currentConverter, //
			Supplier<TextConverter> newConverterCtor) {

		String optValue = commandLine//
				.getOptionValue(optName, "true")//
				.toLowerCase();

		switch (optValue) {
		case "true":
			return currentConverter.andThen(newConverterCtor.get());
		case "false":
			return currentConverter;
		default:
			throw new IllegalArgumentException("Invalid " + optName + " value:" + optValue);
		}
	}

	@Override
	public void setIn(Reader in) {
		this.in = in;
	}

	@Override
	public void setOut(Writer out) {
		this.out = out;
	}

	@Override
	public void setCommandLine(CommandLine commandLine) {
		this.commandLine = commandLine;
	}

	@Override
	public Option[] getOptions() {
		Option[] options = { //
				createTrimCliOption("l", LEFT_TRIM, "trim leading whitespaces"), //
				createTrimCliOption("r", RIGHT_TRIM, "trim trailing whitespaces")//
		};
		return options;
	}

	private Option createTrimCliOption(String shortName, String longName, String description) {
		return Option//
				.builder(shortName)//
				.longOpt(longName)//
				.hasArg(true)//
				.argName("true|false")//
				.required(false)//
				.desc(description + " (default is 'true')")//
				.build();
	}

	private static class LeftTrimmer implements TextConverter {
		boolean trimming = true;

		@Override
		public void convertChar(int charCode, CharWriter w) throws IOException {
			if (!trimming) {
				w.write(charCode);
			} else if (!Character.isWhitespace(charCode)) {
				w.write(charCode);
				trimming = false;
			}

		}
	}

	private static class RightTrimmer implements TextConverter {

		private final Queue<Integer> buffer = new LinkedList<>();

		@Override
		public void convertChar(int charCode, CharWriter w) throws IOException {
			if (Character.isWhitespace(charCode)) {
				buffer.offer(charCode);
			} else {
				flush(w);
				w.write(charCode);
			}
		}

		private void flush(CharWriter w) throws IOException {
			Integer fromBuff;
			while ((fromBuff = buffer.poll()) != null) {
				w.write(fromBuff);
			}
		}
	}

	@FunctionalInterface
	private interface CharWriter {
		void write(int charCode) throws IOException;
	}

	@FunctionalInterface
	private interface TextConverter {

		void convertChar(int charCode, CharWriter w) throws IOException;

		default TextConverter andThen(TextConverter nextConverter) {
			return (charCode, w) -> //
			convertChar(charCode, charCodeConverted -> //
			nextConverter.convertChar(charCodeConverted, w));
		}

		static TextConverter orig() {
			return (charCode, w) -> w.write(charCode);
		}

	}

}

File: src/main/java/io/github/saneea/dvh/feature/text/Trim.kt
package io.github.saneea.dvh.feature.text

import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.FeatureContext
import org.apache.commons.cli.CommandLine
import org.apache.commons.cli.Option
import java.io.Reader
import java.io.Writer
import java.util.*

private const val LEFT_TRIM = "leftTrim"
private const val RIGHT_TRIM = "rightTrim"

private typealias WriteCharFunc = (Int) -> Unit

private typealias ConvertFunc = (Int, WriteCharFunc) -> Unit

private val NO_CONVERSION: ConvertFunc = { charCode, writeFun -> writeFun(charCode) }

private infix fun ConvertFunc.and(nextConverter: ConvertFunc): ConvertFunc =
    { charCode, writeFun -> this(charCode) { nextConverter(it, writeFun) } }

class Trim :
    Feature,
    Feature.CLI,
    Feature.CLI.Options,
    Feature.In.Text.Reader,
    Feature.Out.Text.Writer {

    override lateinit var context: FeatureContext
    override lateinit var inTextReader: Reader
    override lateinit var outTextWriter: Writer
    override lateinit var commandLine: CommandLine

    override val meta = Meta("trim leading and/or trailing whitespaces")

    override fun run() {
        val convertFunc = NO_CONVERSION
            .and(LEFT_TRIM, ::leftTrimmer)
            .and(RIGHT_TRIM, ::rightTrimmer)

        var charCode: Int
        while (inTextReader.read().also { charCode = it } != -1) {
            convertFunc(charCode, outTextWriter::write)
        }
    }

    private fun ConvertFunc.and(
        optName: String,
        newTrimmer: () -> ConvertFunc
    ): ConvertFunc {
        val optValue = commandLine
            .getOptionValue(optName, "true")
            .toLowerCase()
        return when (optValue) {
            "false" -> this
            "true" -> this and newTrimmer()
            else -> throw IllegalArgumentException("Invalid $optName value: $optValue")
        }
    }

    override val options: Array<Option>
        get() {
            return arrayOf(
                createTrimCliOption("l", LEFT_TRIM, "trim leading whitespaces"),
                createTrimCliOption("r", RIGHT_TRIM, "trim trailing whitespaces")
            )
        }

    private fun createTrimCliOption(shortName: String, longName: String, description: String): Option =
        Option
            .builder(shortName)
            .longOpt(longName)
            .hasArg(true)
            .argName("true|false")
            .required(false)
            .desc("$description (default is 'true')")
            .build()

    private fun leftTrimmer(): ConvertFunc {
        var trimming = true

        return { charCode, writeFun ->
            if (!trimming) {
                writeFun(charCode)
            } else if (!Character.isWhitespace(charCode)) {
                writeFun(charCode)
                trimming = false
            }
        }
    }

    private fun rightTrimmer(): ConvertFunc {
        val buffer: Queue<Int> = LinkedList()

        fun flush(writeFun: WriteCharFunc) {
            while (true) {
                val fromBuff = buffer.poll() ?: break
                writeFun(fromBuff)
            }
        }

        return { charCode, writeFun ->
            if (Character.isWhitespace(charCode)) {
                buffer.offer(charCode)
            } else {
                flush(writeFun)
                writeFun(charCode)
            }
        }
    }
}